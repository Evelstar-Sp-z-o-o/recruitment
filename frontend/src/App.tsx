import { FC } from 'react';

import Root from '@/src/pages/Root';
import AppProviders from '@/src/utils/providers/AppProviders';

import './styles/main.scss';

const App: FC = () => {
  return (
    <AppProviders>
      <Root />
    </AppProviders>
  );
};

export default App;
