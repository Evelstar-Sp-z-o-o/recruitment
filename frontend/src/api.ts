import type { Post } from './types';

export const BASE_API_URL = 'http://localhost:3000/api';

const getPosts = async () => {
  const ENDPOINT = '/posts';

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Posts list.');
  }

  const data = await res.json();

  return data as unknown as Post[];
};

const getSinglePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Post details.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const deletePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'DELETE',
  });

  if (!res.ok) {
    throw new Error('Something went wrong when deleting post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const updatePost = async (postId: string, postData: Partial<Post>) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when updating post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const createPost = async (postData: Partial<Post>) => {
  const ENDPOINT = `/posts`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when creating post.');
  }

  const data = await res.json();

  return data as unknown as Post;
};

export { getPosts, getSinglePost, updatePost, createPost, deletePost };
