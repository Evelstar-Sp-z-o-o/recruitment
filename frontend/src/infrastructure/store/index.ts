import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
