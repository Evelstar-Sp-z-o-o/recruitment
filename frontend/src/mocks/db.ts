import { faker } from '@faker-js/faker';
import { factory, primaryKey } from '@mswjs/data';

faker.seed(123);

export const db = factory({
  post: {
    data: {
      body: () => faker.lorem.paragraphs({ min: 1, max: 6 }),
      author: () => faker.internet.email(),
      created: () => faker.date.past().getTime(),
      edited: () => faker.date.recent().getTime(),
      postId: () => faker.string.uuid(),
    },
    id: primaryKey(() => faker.number.int()),
  },
});

let counter = 1;

const users = [];

const createUsers = () => {
  for (let i = 0; i < faker.number.int({ min: 2, max: 10 }); i += 1) {
    users.push(faker.internet.email());
  }
};

createUsers();

const createPosts = () => {
  for (let i = 0; i < faker.number.int({ min: 15, max: 100 }); i += 1) {
    db.post.create({
      data: {
        author: users[faker.number.int({ min: 0, max: users.length - 1 })],
      },
      id: counter,
    });
    counter += 1;
  }
};

createPosts();
