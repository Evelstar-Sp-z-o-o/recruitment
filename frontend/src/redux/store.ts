import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
