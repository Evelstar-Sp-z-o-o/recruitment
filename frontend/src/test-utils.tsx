import { ReactNode } from 'react';

import AppProviders from '@/src/utils/providers/AppProviders';
import { render } from '@testing-library/react';

const AllTheProviders = ({ children }: { children: ReactNode }) => {
  return <AppProviders>{children}</AppProviders>;
};

const customRender = (ui, options?) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
