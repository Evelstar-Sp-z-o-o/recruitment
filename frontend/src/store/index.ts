import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
