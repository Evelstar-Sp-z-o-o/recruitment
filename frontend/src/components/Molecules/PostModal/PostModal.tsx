import { FC, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import AlertDialog from '@/src/components/Atoms/AlertDialog/AlertDialog';
import LoginDialog from '@/src/components/Molecules/LoginDialog/LoginDialog';
import { setUser, useCreatePostMutation, useGetPostsQuery, useUpdatePostMutation, RootState, IPost } from '@/src/store';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Container from '@mui/material/Container';
import Modal from '@mui/material/Modal';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';

const StyledCreatePost = {
  width: '80%',
  minWidth: '300px',
  height: '80%',
  bgcolor: 'background.paper',
};

interface IPostModalProps {
  open: boolean;
  close: () => void;
  response: () => void;
  initialPost?: IPost | null;
  postsCount: number;
}

const PostModal: FC<IPostModalProps> = ({ open, close, response, initialPost, postsCount }) => {
  const [createPost, { error, isSuccess }] = useCreatePostMutation();
  const [updatePost, { error: updateError, isSuccess: updateIsSuccess }] = useUpdatePostMutation();
  const user = useSelector<RootState>((state) => state.user);
  const { t } = useTranslation();
  const dispatch = useDispatch();

  const [openLogin, setOpenLogin] = useState<boolean>(false);
  const [isEmailCorrect, setIsEmailCorrect] = useState<boolean>(true);
  const [isPostCorrect, setIsPostCorrect] = useState<boolean>(true);
  const [isSubmitEnabled, setIsSubmitEnabled] = useState<boolean>(false);
  const [isWrongAuthor, setIsWrongAuthor] = useState<boolean>(false);

  useEffect(() => {
    if (error || isSuccess || updateError || updateIsSuccess) {
      if (error || isSuccess) {
        response(error ?? isSuccess);
      }

      if (updateError || updateIsSuccess) {
        response(updateError ?? updateIsSuccess);
      }
    }
  }, [error, isSuccess, updateError, updateIsSuccess]);

  const handleClose = () => {
    setOpenLogin(false);
  };

  const handleAlertClose = () => {
    close();
    setIsWrongAuthor(false);
  };

  const handleLogin = (e) => {
    e.preventDefault();
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const formData = new FormData(e.currentTarget);
    const formJson = Object.fromEntries((formData as any).entries());
    const email = formJson.email;
    if (!emailRegex.test(email)) {
      setIsEmailCorrect(false);
      return;
    }
    localStorage.removeItem('wasHomeModalShown');
    sessionStorage.setItem('user', email);
    handleClose();

    return dispatch(setUser(email));
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (!user) {
      setOpenLogin(true);

      return;
    }

    if (initialPost && user !== initialPost.data.author) {
      setIsWrongAuthor(true);

      return;
    }

    const formData = new FormData(e.target);
    const formJson = Object.fromEntries((formData as any).entries());
    const body = formJson.body;

    if (!body) {
      setIsPostCorrect(false);

      return;
    }

    const postBody = {
      data: {
        body,
        author: user,
        created: initialPost ? initialPost.data.created : null,
      },
      id: initialPost ? initialPost.id : postsCount + 1,
    };

    initialPost ? updatePost(postBody) : createPost(postBody);
    close();
  };

  const handleChange = (e) => {
    if (!initialPost) {
      setIsSubmitEnabled(!!e.target.value);
    } else {
      setIsSubmitEnabled(e.target.value !== initialPost.data.body && !!e.target.value);
    }
  };

  return (
    <>
      <Modal open={open}>
        <Box className="modal" sx={StyledCreatePost}>
          <Typography variant="h3" component="h2" color="textSecondary">
            {t('create.header')}
          </Typography>
          <Container maxWidth="xl" component="form" sx={{ mt: 2 }} onSubmit={handleSubmit}>
            <TextField
              multiline
              autoFocus
              required
              onChange={handleChange}
              error={!isPostCorrect}
              id="body"
              name="body"
              sx={{ width: '100%' }}
              variant="standard"
              label={t('create.inputLabel')}
              defaultValue={initialPost ? initialPost.data.body : null}
            />
            <Container sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', mt: 3 }}>
              <Button variant="contained" color="error" onClick={close}>
                {t('create.button.cancel')}
              </Button>
              <Button variant="contained" color="success" type="submit" disabled={!isSubmitEnabled}>
                {t('create.button.send')}
              </Button>
            </Container>
          </Container>
        </Box>
      </Modal>

      <LoginDialog open={openLogin} closeDialog={handleClose} handleLogin={handleLogin} isError={!isEmailCorrect} />

      <AlertDialog
        open={isWrongAuthor}
        closeDialog={handleAlertClose}
        label={t('create.button.close')}
        title={t('create.authorEdit')}
      />
    </>
  );
};

export default PostModal;
