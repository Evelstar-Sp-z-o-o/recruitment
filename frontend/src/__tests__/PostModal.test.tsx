import React from 'react';

import PostModal from '@/src/components/Molecules/PostModal/PostModal';
import { db } from '@/src/mocks/db';
import { setUser, store } from '@/src/store';
import { screen, fireEvent, waitFor, render } from '@/src/test-utils';
import '@testing-library/jest-dom';

const postsCount = db.post.count();

describe('PostModal', () => {
  vi.mock('react-i18next', async (importOriginal) => {
    const actual = await importOriginal();
    return {
      ...actual,
      useTranslation: () => ({
        t: (key: string) => key,
      }),
      I18nextProvider: ({ children }) => <div>{children}</div>,
    };
  });

  const close = vi.fn();
  const response = vi.fn();

  it('renders correctly when open', () => {
    render(<PostModal open={true} close={close} response={response} initialPost={null} postsCount={postsCount} />);

    expect(screen.getByText('create.header')).toBeInTheDocument();
    expect(screen.getByText('create.inputLabel')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /create.button.cancel/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /create.button.send/i })).toBeInTheDocument();
  });

  it('handles form submission with valid post content', async () => {
    store.dispatch(setUser('test@example.com'));

    render(<PostModal open={true} close={close} response={response} initialPost={null} postsCount={postsCount} />);

    const postInput = screen.getByRole('textbox', { name: /create\.inputLabel/i });

    fireEvent.change(postInput, { target: { value: 'Test post content' } });

    const sendButton = screen.getByRole('button', { name: /create.button.send/i });

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(close).toHaveBeenCalled();
    });
  });

  it('shows login dialog when user is not logged in and form is submitted', async () => {
    store.dispatch(setUser(null));
    render(<PostModal open={true} close={close} response={response} />);

    const postInput = screen.getByRole('textbox', { name: /create\.inputLabel/i });

    fireEvent.change(postInput, { target: { value: 'Test post content' } });

    const sendButton = screen.getByRole('button', { name: /create.button.send/i });

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(screen.getByText('login.dialog.title')).toBeInTheDocument();
    });
  });

  it('handles closing login dialog and reopening post modal', async () => {
    render(<PostModal open={true} close={close} response={response} />);

    const postInput = screen.getByRole('textbox', { name: /create\.inputLabel/i });

    fireEvent.change(postInput, { target: { value: 'Test post content' } });

    const sendButton = screen.getByRole('button', { name: /create.button.send/i });

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(screen.getByText('login.dialog.title')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByRole('button', { name: /login.dialog.button.cancel/i }));
    await waitFor(() => {
      expect(screen.queryByText('login.dialog.title')).not.toBeInTheDocument();
    });

    expect(screen.getByText('create.header')).toBeInTheDocument();
  });

  it('renders and handles alert dialog', async () => {
    store.dispatch(setUser('test@example.com'));
    render(
      <PostModal
        open={true}
        close={close}
        response={response}
        initialPost={{
          data: {
            author: 'Edgardo71@hotmail.com',
            body: 'Textor tabella carcer somnus delego vestrum dapifer tui. Suspendo animus patria est solitudo decumbo. Somnus solitudo inflammatio denuncio acsi decerno addo tergo.\nAbduco suffragium amitto vitiosus incidunt utor demulceo delibero. Quidem vigilo defluo teres. Consequuntur cotidie temptatio amet.\nSophismata libero aureus patria demo ater volaticus turba. Verto argentum asperiores. Tego vilis volutabrum ver votum.\nTertius vomica curtus tandem. Harum careo vicinus. Utrum vacuus crinis.',
            created: 1704510700018,
            edited: 1721943708413,
            postId: '8a6502a7-2485-48fb-974a-658f5a53213f',
          },
          id: 1,
        }}
      />,
    );

    const postInput = screen.getByRole('textbox', { name: /create\.inputLabel/i });

    fireEvent.change(postInput, { target: { value: 'Test post content' } });

    const sendButton = screen.getByRole('button', { name: /create.button.send/i });

    fireEvent.click(sendButton);

    await waitFor(() => {
      expect(screen.getByText('create.authorEdit')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByRole('button', { name: /create.button.close/i }));
    await waitFor(() => {
      expect(screen.queryByText('create.authorEdit')).not.toBeInTheDocument();
    });
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<PostModal open={true} close={close} response={response} initialPost={null} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
