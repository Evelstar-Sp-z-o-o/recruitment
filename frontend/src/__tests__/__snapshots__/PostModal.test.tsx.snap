// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostModal > should match snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
